plenum/server/message_handlers.py

```diff
@@ -124,123 +148,172 @@ def processor(self, validated_msg: ConsistencyProof, params: Dict[str, Any], frm
         self.node.ledgerManager.processConsistencyProof(validated_msg, frm=frm)
 
 
-class PreprepareHandler(BaseHandler):
+class PropagateHandler(BaseHandler):
     fields = {
-        'inst_id': f.INST_ID.nm,
-        'view_no': f.VIEW_NO.nm,
-        'pp_seq_no': f.PP_SEQ_NO.nm
+        'digest': f.DIGEST.nm
     }
 
+    def __init__(self, node):
+        super().__init__()
+        self.node = node
+
     def validate(self, **kwargs) -> bool:
-        return kwargs['inst_id'] in self.node.replicas.keys() and \
-            kwargs['view_no'] == self.node.viewNo and \
-            isinstance(kwargs['pp_seq_no'], int) and \
-            kwargs['pp_seq_no'] > 0
-
-    def create(self, msg: Dict, **kwargs) -> Optional[PrePrepare]:
-        pp = PrePrepare(**msg)
-        if pp.instId != kwargs['inst_id'] \
-                or pp.viewNo != kwargs['view_no'] \
-                or pp.ppSeqNo != kwargs['pp_seq_no']:
+        return kwargs['digest'] is not None
+
+    def create(self, msg: Dict, **kwargs) -> Propagate:
+        ppg = Propagate(**msg)
+        request = TxnUtilConfig.client_request_class(**ppg.request)
+        if request.digest != kwargs['digest']:
             raise MismatchedMessageReplyException
-        return pp
+        return ppg
 
-    def requestor(self, params: Dict[str, Any]) -> Optional[PrePrepare]:
-        return self.node.replicas[params['inst_id']].get_sent_preprepare(params['view_no'], params['pp_seq_no'])
+    def requestor(self, params: Dict[str, Any]) -> Optional[Propagate]:
+        req_key = params[f.DIGEST.nm]
+        if req_key in self.node.requests and self.node.requests[req_key].finalised:
+            sender_client = self.node.requestSender.get(req_key)
+            req = self.node.requests[req_key].finalised
+            return self.node.createPropagate(req, sender_client)
+        return None
 
-    def processor(self, validated_msg: PrePrepare, params: Dict[str, Any], frm: str) -> None:
-        inst_id = params['inst_id']
-        frm = replica.Replica.generateName(frm, inst_id)
-        self.node.replicas[inst_id].process_requested_pre_prepare(validated_msg,
-                                                                  sender=frm)
+    def processor(self, validated_msg: Propagate, params: Dict[str, Any], frm: str) -> None:
+        self.node.processPropagate(validated_msg, frm)
 
 
-class PrepareHandler(BaseHandler):
+class ThreePhaseMessagesHandler(BaseHandler, metaclass=ABCMeta):
```

Comment:
Do we really need to use handlers for processing of 3PC message requests and responses? Maybe move (copy-paste-like approach) to `MessageReq3pcService `?